### PR TITLE
Ampers: Nora Peters

### PR DESCRIPTION
**Pull request by npeters5**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | One design decision I had to make was how to effectively check and verify the input coming into the Trip initialize method. I chose to write a number of private helper methods that were automatically called during the initialization, and I believe this helped me avoid certain issues later on.
Describe any examples of composition that you encountered in this project, if any | Composition is a has_a relationship between classes, and the code we were given had a number of composition relationships, including the TripDispatcher class which has passengers, divers, and trips; the passenger class has drivers and trips, and the driver class has passengers and trips. 
Describe a _nominal test_ that you wrote for this assignment. |The test where I assert that the start_time for a new trip must be the current time. Without TimeCop its hard to test this exactly, so I just tested that the resulting value was an instance of Time.
Describe an _edge case test_ that you wrote for this assignment. | I believe I'm not quite good at edge cases yet; however in my driver_spec I did account for drivers that previously had zero trips assigned to them.
Describe a concept that you gained more clarity on as you worked on this assignment. | I gained a lot more clarity on the concept SRP and writing helper methods to try and achieve SRP. I also feel I improved on TDD and became more acquainted with the Time class.

### CODE REVIEW

**Comment by tildeee** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | x
Answer comprehension questions | :sparkles:
**Wave 1** |
`Trip` has a helper method to calculate duration | x
`Passenger` has a method to calculate total cost of all trips | x
`Driver` has a method to calculate total revenue | x
Complete tests | x, nominal and few edge cases
**Wave 2** |
`TripDispatcher` has a new method to create trips | x
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | x
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | not tested
Complete tests | x
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | x
Appropriate use of Ruby's `Date` | x
Appropriate helper methods were made to help with complex logic | x
Complete tests | not quite, almost there!
**Overall** | 

Super quickly: your test had a stray `binding.pry` call in the driver_spec!

Overall, nice work! I think your code is largely correct, and is just a little lacking when it comes to some detailed things, like testing thoroughly. You tested the nominal cases really thoroughly: are new trips set correctly, with the appropriate nil values? You have tests for that! But not quite for some edge cases. I know you're working on testing edge cases though :)

Here are a few more ideas on edge cases to test:
- `Trip` `duration` method for start date == end date
- `Passenger` `total cost` method for passengers with no trips
- `Passenger` `all time spent` method for passengers with no trips
- `Driver` methods for trips with a `nil` end_time (you got `Passenger` ones)

I think your helper methods you set up in `Driver` DID set you up for success! Great job! Fantastic!

Quick thing though: in `Driver`'s method `driver_with_oldest_trips`... it checks if the last trip in the driver's `@trips` array has an `end_time` of nil. How are you sure of the order of the trips in `@trips`? What if the trip with an `end_time` of `nil` is in the middle of the array? Also, your implementation of `TripDispatcher#driver_with_oldest_trip` method relies on the driver's trips array being sorted by `end_time`, too.

You didn't have a test for these cases so I don't think you could've been sure :(

You don't have a test for repeating `dispatcher.request_trip` a few times, so even though I know that it will correctly choose the driver named Minnie Dach first because Minnie Dach has no trips, I don't have a test to ensure that it will choose the correct driver amongst drivers who have many trips.

You got the tests in a lot of the enumerables stuff and implemented some sweet enumerables logic. I think that as the complexity of the logic increased, you accidentally relied less on tests, particularly for the logic of assigning a driver.

File: lib/driver.rb [hunk 1]
```diff
@@ -43,5 +44,67 @@ def add_trip(trip)
...
+      end
+      total_hours = total_seconds / 3600
+      return total_hours.round(2)
+    end
+
+    def avg_revenue_per_hour
+      if last_trip_not_nil?
+        avg_rev_per_hour = get_total_revenue / total_hours_driving
+      end
+      return avg_rev_per_hour.round(2)
```
Comment (tildeee) [inline lib/driver.rb:75]: When you run `rake` to run the tests, you end up getting an error saying that `avg_rev_per_hour` is `nil`. That's because there is no value assigned to `avg_rev_per_hour`... there is a value assigned to it IF `last_trip_not_nil?` evaluates to true. Otherwise it's nil!
